File: src/lib/secrets.ts
import { isTauriRuntime } from "./runtime";

const SERVICE = "gene-app";
const ACCOUNT = "openrouter";
const LS_KEY = "gene-app:apiKey";

type InvokeFn = <T>(cmd: string, args?: Record<string, unknown>) => Promise<T>;

let cachedInvoke: InvokeFn | null = null;

async function tauriInvoke(): Promise<InvokeFn> {
  if (cachedInvoke) return cachedInvoke;
  const mod = await import("@tauri-apps/api/core");
  cachedInvoke = mod.invoke as InvokeFn;
  return cachedInvoke;
}

export async function getApiKey(): Promise<string | null> {
  if (isTauriRuntime()) {
    try {
      const invoke = await tauriInvoke();
      const value = await invoke<string | null>("secret_get", {
        service: SERVICE,
        account: ACCOUNT,
      });
      return value && value.length > 0 ? value : null;
    } catch (error) {
      console.error("keyring get failed:", error);
      return null;
    }
  }

  try {
    const value = window.localStorage.getItem(LS_KEY);
    return value && value.length > 0 ? value : null;
  } catch {
    return null;
  }
}

export async function setApiKey(value: string): Promise<void> {
  if (isTauriRuntime()) {
    const invoke = await tauriInvoke();
    await invoke<void>("secret_set", {
      service: SERVICE,
      account: ACCOUNT,
      password: value,
    });
    return;
  }

  try {
    window.localStorage.setItem(LS_KEY, value);
  } catch (error) {
    console.error("localStorage set failed:", error);
  }
}

export async function deleteApiKey(): Promise<void> {
  if (isTauriRuntime()) {
    const invoke = await tauriInvoke();
    await invoke<void>("secret_delete", {
      service: SERVICE,
      account: ACCOUNT,
    });
    return;
  }

  try {
    window.localStorage.removeItem(LS_KEY);
  } catch (error) {
    console.error("localStorage delete failed:", error);
  }
}

File: src-tauri/src/secrets.rs
use keyring::Entry;

#[tauri::command]
pub fn secret_get(service: String, account: String) -> Result<Option<String>, String> {
  let entry = Entry::new(&service, &account).map_err(|e| e.to_string())?;
  match entry.get_password() {
    Ok(p) => Ok(Some(p)),
    Err(keyring::Error::NoEntry) => Ok(None),
    Err(e) => Err(e.to_string()),
  }
}

#[tauri::command]
pub fn secret_set(service: String, account: String, password: String) -> Result<(), String> {
  let entry = Entry::new(&service, &account).map_err(|e| e.to_string())?;
  entry.set_password(&password).map_err(|e| e.to_string())
}

#[tauri::command]
pub fn secret_delete(service: String, account: String) -> Result<(), String> {
  let entry = Entry::new(&service, &account).map_err(|e| e.to_string())?;
  match entry.delete_credential() {
    Ok(_) => Ok(()),
    Err(keyring::Error::NoEntry) => Ok(()),
    Err(e) => Err(e.to_string()),
  }
}
